### PR TITLE
in-memory _source search

### PR DESCRIPTION
**Pull request by fil512**

### CODE REVIEW

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/api/RestSearchParameterTypeEnum.java [hunk 1]
```diff
@@ -89,13 +89,18 @@
...
 	 * Code Value: <b>number</b>
 	 *
 	 * Search parameter SHALL be a number (a whole number, or a decimal).
 	 */
 	SPECIAL("special", "http://hl7.org/fhir/search-param-type"),
 
+	/**
+	 * _source parameter
+	 */
+	SOURCE("string", "http://hl7.org/fhir/search-param-type"),
```
Comment (jamesagnew) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/api/RestSearchParameterTypeEnum.java]: The `_source` parameter is defined in the FHIR spec as being a token. You don't need to define anything new for it, just use it as a token.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/SearchBuilder.java [hunk 2]
```diff
@@ -861,19 +861,10 @@ private Predicate addPredicateSource(List<? extends IQueryParameterType> theList
...
-			if (lastHashValueIndex == -1) {
-				sourceUri = nextParamValue;
-				requestId = null;
-			} else {
-				sourceUri = nextParamValue.substring(0, lastHashValueIndex);
-				requestId = nextParamValue.substring(lastHashValueIndex + 1);
-			}
-			requestId = left(requestId, Constants.REQUEST_ID_LENGTH);
-
+			// FIXME KHS this works, but is it right?
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/SearchBuilder.java]: Well... You are gonna need to parse the token value into the system and transaction id parts.. but I don't think it needs a new class for that. Have a look in `SearchBuilder`, specifically searching for `Constants.PARAM_SOURCE` to see how the regular DB search deals with this param. You could even refactor that parsing bit out of `addPredicateSource(...)` and into its own utility method so you could share.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/api/RestSearchParameterTypeEnum.java (hunk 1 above)
Comment (fil512) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/api/RestSearchParameterTypeEnum.java]: fixed

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/SearchBuilder.java (hunk 2 above)
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/SearchBuilder.java]: fixed

**Pull request merged**